### PR TITLE
Added CNN for centerline detection, brain detection and added possibility for 3d CNN kernel

### LINKED ISSUE #1588: sct_deepseg_sc - Improvements
- [ ] for finale verbose output (syntax to fslview): do not use absolute path for output file. Use relative.
- [ ] add a check to make sure data is 3d (or 2d? could it work on 2d data)

### LINKED ISSUE #1634: Issues with segmentation results
Segmentation fails in the following datasets under: `duke:/sct_testing/large/`:
All: please add tricky cases and list them below for Charley:

State of SCT: commit 60b9b7c366521f852da20ae17a9bd8f10022d32e:

### t2

- zurich_inspired_004/t2

### t2s

- zurich_inspired_004/t2s
- sct_testing_data/mt/mt1.nii.gz
- sct_testing_data/unf_sct_019/t2s/t2s.nii.gz

### dwi

- zurich_inspired_004/dmri
- sct_testing_data/unf_sct_019/dmri/dwi_moco_mean.nii.gz

### LINKED ISSUE #1659: Test DeepSeg on SCI patients from UCSF
Testing new function `sct_deepseg_sc` on acute trauma patients from UCSF (42 SCI patients).
--> tested with SCT v3.1.1

- One issue when trying to segment the full image (see screenshot), issue does not happen when segmenting an image cropped around SC
error_not_crop

- Really good results overall (see below) : 
      - DC against manual segmentations  (mean +- std [min, max]) = 0.84+-0.09 [0.39, 0.93] 
      - When removing the min and max value --> 0.86+-0.07 [0.60, 0.92]

- qualitatively: some leaks and imperfections at the level of compression

- only one patient did really poorly --> strong compressions and change of contrast around SC (patient 1002)

``` 
         DC patient
0   0.900318    1050
1   0.793758    1035
2   0.902963    1052
3   0.892740    1053
4   0.883960    1055
5   0.895695    1056
6   0.920094    1059
7   0.902180    1038
8   0.867812    1039
9   0.781832    1018
10  0.830535    1014
11  0.885860    1015
12  0.874479    1030
13  0.914643    1017
14  0.884393    1010
15  0.922708    1011
16  0.866077    1034
17  0.708985    1013
18  0.912033    1043
19  0.848867    1042
20  0.822940    1041
21  0.656201    1032
22  0.926243    1026
23  0.895595    1033
24  0.880544    1016
25  0.868875    1025
26  0.918082    1024
27  0.839553    1027
28  0.892296    1031
29  0.869572    1047
30  0.798049    1046
31  0.868768    1045
32  0.879650    1022
33  0.866679    1061
34  0.896972    1036
35  0.861692    1049
36  0.867824    1048
37  0.912950    1028
38  0.873283    1037
39  0.876548    1051
40  0.789824    1060
41  0.879428    1007
42  0.779304    1006
43  0.601010    1004
44  0.911359    1003
45  0.388373    1002
46  0.849532    1001
```

### LINKED ISSUE #1716: Quasi-systematic under-segmentation with sct_deepseg_sc
### Description

I've noticed several cord under-segmentations on T1w contrast with `sct_deepseg_sc` when compared to `sct_propseg`. Example below with `sct_019_t1`:

screen shot 2018-05-23 at 12 49 05

Note that I don't observe as much under-segmentation on T2-like contrasts

### PR DESCRIPTION
**Pull request by charleygros**
### TODO

* [x] add more comments / documentation in the script
* [ ] improve terminal output
* [x] put models on OSF --> 599a1bc9f34907e88a6f554aa3372f2dddddf89d
* [ ] comparison `master` vs. `cg_ctr_seg_lesion` branch
* [ ] output the centerline
* [x] in light of the observed results, it was decided to keep the current 2d model, and add to this PR the CNN_1 (detection) and 3d kernels. Regarding the leaking on the t1 and dwi, the ad-hoc solution is to adjust (i.e. lower) the threshold on the inference, as described [here](https://github.com/neuropoly/spinalcordtoolbox/issues/1716#issuecomment-391213046). More investigations will be done to further optimize this threshold.
* [ ] output XX_centerline.nii.gz

### Description of the Change

Regarding the sub-segmentation on the T1 and DWI contrasts, the threshold of the Dice loss used during the training was a sensitive parameters to this issue. In order to find the threshold that produces the best output segmentation, the hyper-parameter optimization has now included the threshold value (range: [0,1], step: 10^-3).

To address the spurious "blobs", a post-processing algorithm `_remove_blobs()` finds the slices with more than one object, then gets the center of mass of each object per "bad" slice, then looks at the above and bottom slice and keep the object with closest center of mass.

- **CENTERLINE DETECTION**: 2 available ways to do it --> via the flag `-ctr`
1. `-ctr svm` --> OptiC initialized with `SVM+HoG` maps (former version, currently the one integrated in `master`). Avantages: faster, more robust to exotic / new contrasts.
2.  `-ctr cnn` --> OptiC initialized with `CNN_1` maps. Avantages: more robust to compressed cord and/or sagittal images.

- **DEALING WITH THE PRESENCE OF THE BRAIN**:
1. `-brain 0` --> The algorithm will go through all the axial slices. Currently the version running in `master`. Note that this option is always set with `-ctr svm`.
2.  `-brain 1` --> If brain is detected (using the CNN1 previously mentioned), the segmentation algorithm will ignore axial slices above the brainstem. PROS: (i) speed up the segmentation (especially for sagittal images), (ii) Less false positive 'blobs' in the brain. CONS: The segmentation could sometimes stop below the brainstem. 

- **CONVOLUTION KERNEL FOR SEGMENTATION**: via the flag `-kernel`
1. `-kernel 2d` --> perform the segmentation of each axial slice one by one. Avantage: Faster + T2 and T2* dataset benefit of a more complex intensity standardization process (slice wise basis) which provided significant during exploratory experiments.
2.  `-kernel 3d` --> use 3D kernels for the segmentation. Could help for compressed cord and/or with lesion and/or sagittal images, but is slower (0.5 iso 3D resampling + use of 3D convolution are responsible for it).

### Steps and Constraints

/!\ Please re-install the toolbox: both `spinalcordtoolbox` and `data` folders

### Related issues

Fixes #1634, Fixes #1716, Fixes #1659, Fixes #1588

### CODE REVIEW

**Comment by jcohenadad** [comment]
Error:
~~~
spine_generic/20171127_douglas/t2s $ sct_deepseg_sc -i t2s.nii.gz -c t2s -ofolder sct-cg_ctr_seg_lesion

--
Spinal Cord Toolbox (cg_ctr_seg_lesion/962f04d46bd68c1276eb2455718952b893405ac4)
Running /Users/julien/code/sct-cnn/scripts/sct_deepseg_sc.py -i t2s.nii.gz -c t2s -ofolder sct-cg_ctr_seg_lesion

Method:
	Centerline algorithm: cnn
	Assumes brain section included in the image: False
	Dimension of the segmentation kernel convolutions: 2d

Creating temporary folder...
cp t2s.nii.gz /var/folders/6f/wy6ljmx9453cgth2qwv5l1s80000gn/T/sct-180606154803-rSl7uJ
Reorient the image to RPI, if necessary...
cp /var/folders/6f/wy6ljmx9453cgth2qwv5l1s80000gn/T/sct-180606154803-rSl7uJ/t2s.nii.gz t2s_RPI.nii.gz
Resample the image to 0.5 mm isotropic resolution...
/Users/julien/code/sct-cnn/python/lib/python2.7/site-packages/nipy/io/files.py:145: FutureWarning: Default `strict` currently False; this will change to True in a future version of nipy
  ni_img = nipy2nifti(img, data_dtype = io_dtype)
Finding the spinal cord centerline...
Traceback (most recent call last):
  File "/Users/julien/code/sct-cnn/scripts/sct_deepseg_sc.py", line 844, in <module>
    main()
  File "/Users/julien/code/sct-cnn/scripts/sct_deepseg_sc.py", line 835, in main
    remove_temp_files=remove_temp_files, verbose=verbose)
  File "/Users/julien/code/sct-cnn/scripts/sct_deepseg_sc.py", line 646, in deep_segmentation_spinalcord
    ctr_model = nn_architecture_ctr(height=dct_patch_ctr[contrast_type]['height'],
KeyError: 'height'
~~~

**Comment by jcohenadad** [comment]
20171127_douglas t1

**Comment by charleygros** [comment]
@jcohenadad : [This bug](https://github.com/neuropoly/spinalcordtoolbox/pull/1822#issuecomment-394959337) should be now fixed by the last commit.

File: scripts/sct_deepseg_sc.py [hunk 1]
```diff
@@ -372,18 +743,19 @@ def deep_segmentation_spinalcord(fname_image, contrast_type, output_folder, remo
 
     # binarize the resampled image to remove interpolation effects
     sct.log.info("Binarizing the segmentation to avoid interpolation effects...")
-    sct.run(['sct_maths', '-i', fname_seg_RPI, '-bin', '0.75', '-o', fname_seg_RPI], verbose=0)
+    thr = '0.000001' if contrast_type in ['t1', 'dwi'] else '0.5'
```
Comment (zougloub) [inline scripts/sct_deepseg_sc.py]: Why 0.000001?

**Comment by jcohenadad** [comment]
Bug with commit 281e11a775e8166f1bb26a5941439904ec5896c4

~~~
Traceback (most recent call last):
  File "/Users/julien/code/sct-cnn/scripts/sct_deepseg_sc.py", line 804, in <module>
    main()
  File "/Users/julien/code/sct-cnn/scripts/sct_deepseg_sc.py", line 795, in main
    remove_temp_files=remove_temp_files, verbose=verbose)
  File "/Users/julien/code/sct-cnn/scripts/sct_deepseg_sc.py", line 662, in deep_segmentation_spinalcord
    fname_out=fname_seg_crop)
  File "/Users/julien/code/sct-cnn/scripts/sct_deepseg_sc.py", line 435, in segment_2d
    seg_model.load_weights(model_fname)
  File "/Users/julien/code/sct-cnn/python/lib/python2.7/site-packages/keras/engine/topology.py", line 2566, in load_weights
    f = h5py.File(filepath, mode='r')
  File "/Users/julien/code/sct-cnn/python/lib/python2.7/site-packages/h5py/_hl/files.py", line 271, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)
  File "/Users/julien/code/sct-cnn/python/lib/python2.7/site-packages/h5py/_hl/files.py", line 101, in make_fid
    fid = h5f.open(name, flags, fapl=fapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/private/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T/pip-wdzlRM-build/h5py/_objects.c:2840)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/private/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T/pip-wdzlRM-build/h5py/_objects.c:2798)
  File "h5py/h5f.pyx", line 78, in h5py.h5f.open (/private/var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T/pip-wdzlRM-build/h5py/h5f.c:2117)
IOError: Unable to open file (Unable to open file: name = '/users/julien/code/sct-cnn/data/deepseg_sc_models/t1_sc.h5', errno = 2, error message = 'no such file or directory', flags = 0, o_flags = 0)
~~~

This error is caused by the naming of the model folder (the one installed under data/) as `deepseg_sc` instead of `deepseg_sc_models`. 
@charleygros i've manually fixed the naming on my local station, but could you please update on OSF? Thanks

**Comment by jcohenadad** [comment]
tested 281e11a775e8166f1bb26a5941439904ec5896c4:

Legend: OK+++: perfect, OK++: leaking detected in 1 slice, OK+: leaking detected on several non-contiguous slices.

- 20171127_douglas/t1 --> OK+++ (issue seen at https://github.com/neuropoly/spinalcordtoolbox/pull/1822#issuecomment-394962931 is fixed)
- 20171127_douglas/t2 --> OK+++
- sct_example_data/t1 --> OK+++ (issue seen at https://github.com/neuropoly/spinalcordtoolbox/pull/1822#issuecomment-394962931 is fixed)
- sct_example_data/mt/t1w --> OK+++ (issue seen at https://github.com/neuropoly/spinalcordtoolbox/pull/1822#issuecomment-394962931 is fixed)
- sct_example_data/t2s --> OK+: segmentation is mostly fine, except at slice 31 where there is not output segmentation despite the contrast being reasonably good.
sct_example_data/mt/mt1 (using -c t2) --> OK+++
- 20171221_poly/t2s --> OK++: a bit of leaking but mostly fine
- 20171207_ucl/t2s --> OK+++
- 20180604_tokyo-univ-ge750w/t2s --> OK+: some leaking in the top slices
- 20171207_ucl/t1 --> OK++: some leaking at slices 59, 79, but fine otherwise.
- 20171221_poly/t1 --> OK++ (slice 113)
- 20180608_chiba-ge/t1 --> OK++ (slice 112)

Great job @charleygros! Should be ready to merge after https://github.com/neuropoly/spinalcordtoolbox/pull/1822#issuecomment-396039091 is addressed.

**Comment by jcohenadad** [comment]
i fixed https://github.com/neuropoly/spinalcordtoolbox/pull/1822#issuecomment-396039091 by uploading new models (both on OSF and neuropoly): 20180610_deepseg_sc_models.zip.

**Comment by jcohenadad** [comment]
found major flaw with cnn for centerline detection on data from SCT course:
red: svm, pink: cnn (i.e. no centerline detected)
screen shot 2018-06-11 at 11 10 09

i suggest we always use svm for centerline detection by default until the cnn is robust enought. @charleygros agreed?

**Pull request merged**